core/src/coproduct.rs

```diff
@@ -144,28 +144,361 @@ macro_rules! Coprod {
+    /// let co1_as_f32: Option<f32> = co1.take();
+    /// assert_eq!(co1_as_i32, None);
+    /// assert_eq!(co1_as_f32, Some(42f32));
+    ///
+    /// // You can also use turbofish syntax to specify the type.
+    /// // The Index parameter should be left as `_`.
+    /// let co2 = I32F32::inject(1i32);
+    /// assert_eq!(co2.take::<i32, _>(), Some(1));
+    /// assert_eq!(co2.take::<f32, _>(), None);
+    /// # }
+    /// ```
+    #[inline(always)]
+    pub fn take<T, Index>(self) -> Option<T>
+    where Self: CoproductTaker<T, Index>,
+    {
+        CoproductTaker::take(self)
+    }
+
+    /// Attempt to extract a value from a coproduct (or get the remaining possibilities).
+    ///
+    /// By chaining calls to this, one can exhaustively match all variants of a coproduct.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// # #[macro_use] extern crate frunk_core;
+    /// # fn main() {
+    /// type I32F32 = Coprod!(i32, f32);
+    /// type I32 = Coprod!(i32); // remainder after uninjecting f32
+    /// type F32 = Coprod!(f32); // remainder after uninjecting i32
+    ///
+    /// let co1 = I32F32::inject(42f32);
+    ///
+    /// // You can let type inference find the desired type.
+    /// let co1 = I32F32::inject(42f32);
+    /// let co1_as_i32: Result<i32, F32> = co1.uninject();
+    /// let co1_as_f32: Result<f32, I32> = co1.uninject();
+    /// assert_eq!(co1_as_i32, Err(F32::inject(42f32)));
+    /// assert_eq!(co1_as_f32, Ok(42f32));
+    ///
+    /// // It is not necessary to annotate the type of the remainder:
+    /// let res: Result<i32, _> = co1.uninject();
+    /// assert!(res.is_err());
+    ///
+    /// // You can also use turbofish syntax to specify the type.
+    /// // The Index parameter should be left as `_`.
+    /// let co2 = I32F32::inject(1i32);
+    /// assert_eq!(co2.uninject::<i32, _>(), Ok(1));
+    /// assert_eq!(co2.uninject::<f32, _>(), Err(I32::inject(1)));
+    /// # }
+    /// ```
+    ///
+    /// Chaining calls for an exhaustive match:
+    ///
+    /// ```rust
+    /// # #[macro_use] extern crate frunk_core;
+    /// # fn main() {
+    /// type I32F32 = Coprod!(i32, f32);
+    ///
+    /// // Be aware that this particular example could be
+    /// // written far more succinctly using `fold`.
+    /// fn handle_i32_f32(co: I32F32) -> f32 {
+    ///     // Remove i32 from the coproduct
+    ///     let co = match co.uninject::<i32, _>() {
+    ///         Ok(x) => return (2 * x) as f32,
+    ///         Err(co) => co,
+    ///     };
+    ///
+    ///     // Remove f32 from the coproduct
+    ///     let co = match co.uninject::<f32, _>() {
+    ///         Ok(x) => return 2.0 * x,
+    ///         Err(co) => co,
+    ///     };
+    ///
+    ///     // Now co is empty
+    ///     match co { /* unreachable */ }
+    /// }
+    ///
+    /// assert_eq!(handle_i32_f32(I32F32::inject(3)), 6.0);
+    /// assert_eq!(handle_i32_f32(I32F32::inject(3.0)), 6.0);
```

Comment:
Nit: I wonder if this test would be more illustrative if the results from `handle_i32_f32` were different when `i32` is pulled out vs `f32` (e.g. one multiplies by 2, the other by 3)